### ISSUE TITLE
metallb and services with no selector (endpoints updates) won't work

### ISSUE THREAD

**Issue by schmitch** [google/metallb#305]
TL;DR;
since 0.7 metallb does not work with services that have no nodeSelector instead updating the endpoints directly.

--

**Environment**:
- MetalLB version: 0.7.2
- Kubernetes version: 1.11.0
- BGP router type/version: Layer2
- OS (e.g. from /etc/os-release): CoreOS 1800.7.0 or 1800.6.0 (nodes updating automatically)
- Kernel (e.g. `uname -a`): 4.14.63-coreos or 4.14.59-coreos-r2

Currently I have a working metallb setup. It works for three http routes, perfectly fine.
Both of them are from the nginx-ingress-controller and one is a cntlm proxy. I assigned them 10.10.30.100 - 10.10.30.102 (so three ip's, my pool goes form 10.10.30.100-10.10.30.200).
All are curlable on the opened port.

However now I created a PostgreSQL deployment via https://github.com/zalando-incubator/postgres-operator and the LB will **never** work.

<b>
Actually the stuff worked on pre 0.7, i.e. were everything was ClusterIP. Currently I can switch to Local or ClusterIP and nothing will happen, so that's not a problem.
And I actually have the same problem in our in house cluster, I'm pretty sure minikube deployments do not work, aswell, but it's pretty easy to reproduce, create a endpoints pointing to a deployment node and create a real service with LoadBalancer, no ip will work.
</b>

I guess the problem is somehow, how postgres-operator works.
It uses the endpoints mechanism to assign the nodes to the Service.

This is how the service looks like:
```
apiVersion: v1
kind: Service
metadata:
  annotations:
    external-dns.alpha.kubernetes.io/hostname: loki-cluster.production.db.example.com
    service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: "3600"
  creationTimestamp: 2018-07-08T14:52:51Z
  labels:
    application: spilo
    spilo-role: master
    version: production-loki-cluster
  name: production-loki-cluster
  namespace: storz-production
  resourceVersion: "19650525"
  selfLink: /api/v1/namespaces/storz-production/services/production-loki-cluster
  uid: 96b9ba61-82be-11e8-8cc6-0050569f2fb1
spec:
  clusterIP: 10.107.249.206
  externalTrafficPolicy: Cluster
  loadBalancerIP: 10.10.30.200
  ports:
  - name: postgresql
    nodePort: 32242
    port: 5432
    protocol: TCP
    targetPort: 5432
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer:
    ingress:
    - ip: 10.10.30.200
```

As you can see, it has no selector, so it won't get the nodes over the selector, instead it uses the endpoints here:
```
apiVersion: v1
kind: Endpoints
metadata:
  annotations:
    acquireTime: 2018-08-10T08:44:04.091844+00:00
    leader: production-loki-cluster-1
    optime: "73613704680"
    renewTime: 2018-08-20T07:39:28.680138+00:00
    transitions: "20"
    ttl: "30"
  creationTimestamp: 2018-07-08T14:52:51Z
  labels:
    application: spilo
    spilo-role: master
    version: production-loki-cluster
  name: production-loki-cluster
  namespace: storz-production
  resourceVersion: "19652490"
  selfLink: /api/v1/namespaces/storz-production/endpoints/production-loki-cluster
  uid: 96a889ab-82be-11e8-8cc6-0050569f2fb1
subsets:
- addresses:
  - ip: 172.31.4.180
  ports:
  - name: postgresql
    port: 5432
    protocol: TCP
```

My guess is that the new release of metallb does not work well with that mechanism.
Currently Spilo/postgres-operator will update the endpoints on failover, so that the service will point to the new ip's.

Maybe that issue is related: https://github.com/google/metallb/issues/302
I mean it would be great if kubernetes could actually route to other services, so that I could have a outer-postgres-service that works well with metallb work with the inner postgres service, but I guess that would make the problem worse.

Edit: of course I can go around the problem by using haproxy -> psql and having haproxy as my LoadBalancer with a selector), but I guess using the Service->Endpoints should work in metallb aswell.

**Comment by danderson** [comment]
Selectors don't matter to MetalLB. As long as the endpoints are updating and contain ready addresses, it should work. So it sounds like it's something else in your cluster that's broken. Unfortunately, as you probably guessed from the fact that it took >6 months for me to reply, I have no time to troubleshoot user clusters, so I'm afraid unless you can figure out a root cause, I can't help you further. You could try asking on the #metallb channel on the kubernetes slack.

**Comment by teralype** [comment]
@schmitch did you find what the problem was?

**Comment by schmitch** [comment]
nope. I couldn't get metallb + zalando postgres operator to work.

**Comment by kalak451** [comment]
In case anyone bumps into this same thing with metallb + zalando,  the thing that is preventing it from working is that zolando is setting loadBalancerSourceRanges on the service to something completely unrelated to your network.  If you directly edit the service and remove this setting or change it to a sane value, everything works great.   Definitely an issue with zalando and not metallb.

**Comment by TheKangaroo** [comment]
as an addition to @kalak451's comment:
you can specify the loadBalancerSourceRanges in your postgres CRD like:
```
```yaml
apiVersion: "acid.zalan.do/v1"
kind: postgresql
  name: pgsql-test
spec:
[...]
  allowedSourceRanges:
    - 0.0.0.0/0
```
And you do not need to edit the service manually :)
